apps/aecore/lib/aecore/sync/sync.ex

```diff
@@ -0,0 +1,843 @@
+               peer_id,
+               top_header_hash,
+               top_block_height,
+               max_agree,
+               max_agree,
+               Genesis.height(),
+               min_agreed_hash
+             ) do
+          {:ok, block_height, header_hash} ->
+            Logger.info(
+              "#{__MODULE__}: Agreed upon height: #{block_height} with #{inspect(peer_id)}"
+            )
+
+            agreement = {:agreed_height, %{height: block_height, hash: header_hash}}
+            do_work_on_sync_task(peer_id, task_id, agreement)
+
+          {:error, reason} ->
+            do_work_on_sync_task(peer_id, task_id, {:error, {:agree_on_height, reason}})
+        end
+
+      {:fill_pool, start_header_hash, target_header_hash} ->
+        fill_pool(peer_id, start_header_hash, target_header_hash, task_id)
+
+      {:post_blocks, blocks} ->
+        res = post_blocks(blocks)
+        do_work_on_sync_task(peer_id, task_id, {:post_blocks, res})
+
+      {:get_block, block_height, header_hash} ->
+        res =
+          case do_fetch_block(header_hash, peer_id) do
+            {:ok, false, _block} ->
+              {:get_block, block_height, header_hash, peer_id, {:ok, :local}}
+
+            {:ok, true, block} ->
+              {:get_block, block_height, header_hash, peer_id, {:ok, block}}
+
+            {:error, reason} ->
+              {:error, {:get_block, reason}}
+          end
+
+        do_work_on_sync_task(peer_id, task_id, res)
+
+      :abort_work ->
+        Logger.info("#{__MODULE__}: #{self()} Aborting sync work against #{inspect(peer_id)}")
+        schedule_ping(peer_id)
+    end
+  end
+
+  defp post_blocks([]), do: :ok
+
+  defp post_blocks([{start_height, _, _} | _] = blocks) do
+    post_blocks(start_height, start_height, blocks)
+  end
+
+  defp post_blocks(from, to, []) do
+    Logger.info("#{__MODULE__}: Synced blocks from #{from} to #{to}")
+    :ok
+  end
+
+  defp post_blocks(from, _to, [{block_height, _header_hash, {_peer_id, :local}} | blocks]) do
+    post_blocks(from, block_height, blocks)
+  end
+
+  defp post_blocks(from, _to, [{block_height, _header_hash, {peer_id, block}} | blocks]) do
+    case Chainstate.add_block(block) do
+      :ok ->
+        post_blocks(from, block_height, blocks)
+
+      {:error, reason} ->
+        Logger.info("#{__MODULE__}: Failed to add synced block #{block_height}: #{reason}")
+        {:error, peer_id, block_height}
+    end
+  end
+
+  defp agree_on_height(_peer_id, _r_header_hash, _rheight, _lheight, min, min, agreed_header_hash) do
+    {:ok, min, agreed_header_hash}
+  end
+
+  ## Ping logic makes sure they always agree on genesis header (height 0)
+  ## We look for the block that is both on remote highest chain and in our local
+  ## chain, connected to genesis (may be on a fork, but that fork
+  ## has now more difficulty than our highest chain (otherwise we would not sync))
+  ## agreed_hash is hash at height min (genesis hash)
+  defp agree_on_height(
```

Comment:
It is only one version of this function. The different bodies of it are because of the guards and pattern-matching. Maybe it is confusing because the explanation of the function is on top of the second body, will move it to the top one.

Reply:
Just the execution flow between the two bodies is a little bit confusing for me, but it's ok to leave it as is.